File: modulars/src/main/java/com/eric/modulars/MainActivity.java
package com.eric.modulars;

import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;

import com.eric.modulars.service.LoginService;
import com.eric.modulars.service.LoginServiceImp;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;

public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        LoginServiceImp imp = new LoginServiceImp();
        imp.login();
        System.out.println("===================");

        // 反射获取
        try {
            Class<?> aClass = Class.forName("com.eric.modulars.service.LoginServiceImp");
            Method method = aClass.getDeclaredMethod("login");
            Object o = aClass.newInstance();
            method.invoke(o);


            // 或者这样
            LoginService service = (LoginService) o;
            service.login();

        } catch (ClassNotFoundException e) {
            e.printStackTrace();
        } catch (NoSuchMethodException e) {
            e.printStackTrace();
        } catch (IllegalAccessException e) {
            e.printStackTrace();
        } catch (InstantiationException e) {
            e.printStackTrace();
        } catch (InvocationTargetException e) {
            e.printStackTrace();
        }
    }
}

File: app/src/main/java/com/eric/jetpack/viewModel/MainActivity.kt
package com.eric.jetpack.viewModel

import android.content.Intent
import android.os.Bundle
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.ViewModelProvider
import com.eric.jetpack.R
import com.eric.jetpack.annotation.AnnotationSample
import com.eric.jetpack.coroutine.CoroutineActivity
import com.eric.jetpack.lifecycles.LifecycleActivity
import com.eric.jetpack.liveData.LiveDataActivity
import com.eric.jetpack.retrofit.RetrofitActivity
import com.eric.jetpack.thread.ThreadActivity
import kotlinx.android.synthetic.main.activity_main.*


class MainActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        val model = ViewModelProvider(this, MainVmodelFactory("2222")).get(MainVModel::class.java)

        tv_text.text = model.text.toString() + "  id:" + model.id

        tv_text.setOnClickListener {
            model.text++
            tv_text.text = model.text.toString()
        }

        initListener()
    }


    private fun initListener() {
        tv_lifeCycles.setOnClickListener {
            val intent = Intent(this, LifecycleActivity::class.java)
            startActivity(intent)
        }

        tv_liveData.setOnClickListener {

            val intent = Intent(this, LiveDataActivity::class.java)
            startActivity(intent)
        }

        tv_retrofit.setOnClickListener {
            val intent = Intent(this, RetrofitActivity::class.java)
            startActivity(intent)
        }

        tv_thread.setOnClickListener {
            val intent = Intent(this, ThreadActivity::class.java)
            startActivity(intent)
        }


        tv_coroutine.setOnClickListener {
            val intent = Intent(this, CoroutineActivity::class.java)
            startActivity(intent)
        }

    }

}